In [4]:
import pandas as pd
import numpy as np
import pickle

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.ensemble import RandomForestRegressor

# Load cleaned dataset
df = pd.read_csv("../../data/processed/student_lifestyle_cleaned.csv")

# Encode stress level
stress_map = {
    "Low": 0,
    "Moderate": 1,
    "High": 2
}
df["Stress_Level"] = df["Stress_Level"].map(stress_map)

# Define features and target
feature_cols = [
    "Study_Hours_Per_Day",
    "Sleep_Hours_Per_Day",
    "Social_Hours_Per_Day",
    "Physical_Activity_Hours_Per_Day",
    "Extracurricular_Hours_Per_Day",
    "Stress_Level"
]

X = df[feature_cols]
y = df["GPA"]

# Shared split for everyone
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Task 2 - Random Forest Regressor

Train a non-linear Random Forest model, apply 5-fold cross validation on the training set, and evaluate its performance on the test set. This model captures complex relationships between features and the target variable.

In [5]:
rf_model = RandomForestRegressor(random_state=42, n_estimators=100)

cv_r2_scores = cross_val_score(
    rf_model,
    X_train,
    y_train,
    cv=5,
    scoring='r2'
)

rf_model.fit(X_train, y_train)
y_pred = rf_model.predict(X_test)

test_r2 = r2_score(y_test, y_pred)
test_mse = mean_squared_error(y_test, y_pred)
test_rmse = np.sqrt(test_mse)

print('CV R² (mean):', cv_r2_scores.mean())
print('Test R²:', test_r2)
print('Test RMSE:', test_rmse)

# Save the model
with open('../../models/random_forest_model.pkl', 'wb') as f:
    pickle.dump(rf_model, f)

print('\nModel saved to ../../models/random_forest_model.pkl')

CV R² (mean): 0.47644146920375113
Test R²: 0.44136852397900883
Test RMSE: 0.22837609643962303

Model saved to ../../models/random_forest_model.pkl


## Short Explanation

Random Forest is a model that uses many decision trees and combines their results to make better predictions. It works well when the relationship between variables is not simple or linear.

The cross-validation R² is 0.47, which shows that the model has moderate and consistent performance across different data splits. The test R² is 0.44, meaning the model explains about 44% of the variation in GPA on unseen data. The RMSE is 0.23, indicating that the prediction error is relatively small. 

The model is also saved so it can be used later for feature importance analysis and comparison with other models.